Add tests for message processing flow

michaelansel opened the pull request.

michaelansel committed:
Add tests for message processing flow

michaelansel committed:
Add tests for Listener

michaelansel committed:
Add tests for Message

michaelansel committed:
Refactor Listener tests

michaelansel committed:
Refactor Robot tests

afeld commented on script/bootstrap:
```diff
@@ -15,3 +15,6 @@ if [ -f .env.${NODE_ENV} ]; then
 npm install
+
+# Make sure mocha and coffee are in the PATH
+export PATH="node_modules/.bin:$PATH"
```
This will stop working after the user closes their terminal, so probably better to use the relative path (`./node_modules/.bin/mocha`, etc.) from the scripts. Good call though!

michaelansel replied:
Is this file not intended to always be sourced immediately prior to running tests? I was working under the assumption you would always run `script/test`, which sources this file.

I think the specific path will break if mocha is installed globally instead of within the local directory. I'll test that out to be sure though.

afeld replied:
Oh, [true](https://github.com/afeld/hubot/blob/6dc8db3288f72beb89f93daa399f0eef8f225db4/script/test#L3). Now realizing we can/should probably get rid of [the Makefile](https://github.com/afeld/hubot/blob/6dc8db3288f72beb89f93daa399f0eef8f225db4/Makefile), since the `scripts` are basically doing the same job.

michaelansel replied:
It looks like `make test` is called by [cibuild](https://github.com/afeld/hubot/blob/testing/script/cibuild), so I just updated the Makefile to call `script/test` instead.

afeld commented on package.json:
```diff
@@ -30,6 +30,7 @@
   },
   "devDependencies": {
     "chai": "*",
+    "hubot-mock-adapter": "*",
```
Since this depends on a released version of Hubot, we wouldn't catch any breaking API changes. Maybe we need to use something like [mockery](https://github.com/mfncooper/mockery), to force [their `require('hubot/src/adapter')`](https://github.com/blalor/hubot-mock-adapter/blob/5cb366c9150693624050bc2b04e39c83470de5e8/index.js#L23) to be relative? Does that make sense?

michaelansel replied:
Great catch! I'll dig into that some more.

michaelansel replied:
Updated. I also ripped out most of the unnecessary creation of Robot objects.

afeld commented on test/message_test.coffee:
```diff
@@ -0,0 +1,41 @@
+    @user = @robot.brain.userForId '1', {
+      name: 'hubottester'
+      room: '#mocha'
+    }
```
Can we minimize the dependencies here and just instantiate a `User` directly? You shouldn't need a `Robot` at all.

michaelansel replied:
Most likely; I'll try that out today.

michaelansel replied:
Done

afeld commented on test/message_test.coffee:
```diff
@@ -0,0 +1,41 @@
+          expect( testMessage.match(/^message123$/) ).to.be.ok
+          expect( testMessage.match(/^does-not-match$/) ).to.not.be.ok
+
+  describe 'Message Processing', ->
```
Mind removing the callback here to show that it's [pending](http://mochajs.org/#pending-tests)?

michaelansel replied:
That was a copy-pasta mistake. It shouldn't even be there.

michaelansel replied:
Removed

afeld commented:
Just a few comments – nice work!

michaelansel committed:
Address PR feedback

- Remove unnecessary dependencies
- Ensure hubot-mock-adapter uses the latest Adapter code

michaelansel committed:
Update Makefile to use script/test

afeld commented:
:rocket:
